main.nf

```diff
@@ -19,24 +19,44 @@ process remove_tech_duplicates {
     """
 }
 
+process build_cell_ontology_dict{
+    publishDir "${params.results_dir}", mode: 'copy'
+    conda "${baseDir}/envs/cell_types_analysis.yaml"
+    input:
+        file("metadata_filtered.tsv") from REF_LABELS_FILTERED
```

Comment:
It gets used in the following processes (get_tool_performance_table and generate_empirical_cdf) via the REF_LABELS_FILTERED channel, but as a variable instead of by the actual file name. Is it Ok?

Reply:
What you're doing in this statement is saying 'rename the input of channel REF_LABELS_FILTERED to the literal filename 'metadata_filtered.tsv'. 

But since the process is built entirely from elements of 'params' the input is not needed anyway.

Reply:
I agree, since you take all parameters from config, you don't need this input. But I guess it would be good practice to create a channel for this parameter before the process, so that it can check for file's existence. That will also make workflow more readable

Reply:
Ok I see! Shall we keep the channel then? 
I'll change the 'metadata_filtered.tsv' for variables also. I added the parameter metadata_dir as publishDir of the remove_duplicates output metadata_filtered.tsv since having two files (this, and input metadata.tsv) was causing error. But it can actually be overwritten and avoid having the filtered file in a different directory.

Reply:
@a-solovyev12 is right- you should put all your files into channels which you pass in to this process, rather than having the process read the config.

Reply:
and you should not output to paths derived from the params.

Reply:
... so instead of :

```
output: 
        file("${params.ontology_dict}") into ONTOLOGY_DICT
```

you should have:

```
output: 
        file(outFile) into ONTOLOGY_DICT
```

... which you pass to the process like:

```
--output-dict-path $outFile
```

Reply:
Ok thanks for the comments guys! I'll apply these changes tomorrow

Reply:
The way to make the files you generate appear outside of the nextflow processes is with 'publishDir'.

Reply:
Actually my outFile thing won't work. Hang on...

Reply:
@pinin4fjords I think it can be just file name, like `file('cl_dictionary.rds') into ONTOLOGY_DICT `

Reply:
Deleted my above comments because I was talking rubbish ;-). You're doing file taking output file names from params as long as it's not actual paths.

Reply:
> @pinin4fjords I think it can be just file name, like `file('cl_dictionary.rds') into ONTOLOGY_DICT `

OK and then remove the config parameter corresponding to this file name.

Reply:
Sorry for confusing matters - obviously I'm not thinking clearly. Taking strings from params is fine @rubenchazarra - disregard my earlier comments.

Reply:
No worries, this are constructive confusions!

Reply:
I was concerned you were outputting directly to paths defined in the params, but I don't think that's what you're doing. But getting input files into channels as Andrey suggested, rather than consuming then directly from params, is still the right thing to do :-)